Implement numba.cross2d

partmor opened the pull request:
As discussed in #4128, it is not possible to cover all the functionality of `np.cross` via `overload` since the implementation depends on both the `dim` (accessible in the typing scope) and `shape` (**not** accessible in the typing scope).

Hence we agreed to manage  the following two cases separately:

+ the 3D case: already implemented in the `np.cross` overload in #4128 
+ the 2D case: covered in this PR.

As discussed in the gitter chat with @stuartarchibald, I will proceed to add the `cross2d` implementation to `numba.targets.arraymath` and make it available through `numba.special`.

I am writing the tests in `numba.tests.test_np_functions`.

stuartarchibald commented:
@partmor thanks for this, what you describe sounds good and its relatively easy to move about as needed once implemented :) Thanks again!

partmor commented:
Hi @stuartarchibald , I have a *little* problem. Just by doing in `special.py`:
```python
from .targets.arraymath import cross2d
```
 I get many circular-import-related errors. It seems like it's mainly because many (sub)modules like `arraymath` and `arrayobj` have absolute imports of the form `from numba import ... ` relying on the imports defined in `numba/__init__.py`.

One thing I thought is creating a wrapper in `special.py` of the form:
```python
def cross2d(**args):
    from .targets.arraymath import cross2d as _cross2d
    return _cross2d(**args)
```
but since `cross2d` is expected to be *njitted*, the import would break the users' code (right?).

The only thing I can think of now, is either:
1. put `cross2d` elsewhere (I don't have a candidate location though)
2. try to replace the absolute `from numba import ...` imports in the affected modules with relative imports that point directly to the *actual* location of the function.

E.g.: in `arraymath`, replace
```python
from numba import generated_jit
```
with
```python
from ..decorators import generated_jit
```
and so on (there are a number of these that would require refactoring - probably this is out of scope for this PR..).

What do you think? Am I missing out on something?

Thanks!

stuartarchibald commented:
RE: https://github.com/numba/numba/pull/4595#issuecomment-533867091, noted, thanks, I can reproduce. Whilst chatting out of band @sklam pointed out that PEP8 prefers absolute imports, perhaps because Python 3 has better circular import resolution ability.

To resolve this reasonably quickly, how about creating a `numba.targets.array_math_ext` module for array math extensions, and expose it from there? Do you think that'd work? I think this is quite cheap to try. Thanks.

partmor commented:
> RE: [#4595 (comment)](https://github.com/numba/numba/pull/4595#issuecomment-533867091), noted, thanks, I can reproduce. Whilst chatting out of band @sklam pointed out that PEP8 prefers absolute imports, perhaps because Python 3 has better circular import resolution ability.
> 
> To resolve this reasonably quickly, how about creating a `numba.targets.array_math_ext` module for array math extensions, and expose it from there? Do you think that'd work? I think this is quite cheap to try. Thanks.

@stuartarchibald  Aha! I didn't know it was actually a PEP8 recommendation. Thanks for the hint.

+ I have implemented a `numba.targets.array_math_ext` module that exposes the `cross2d` function.
+ I have updated the documentation to reference `cross2d`.
+ I have updated the exception message in the `np.cross` overload to hint the usage of `cross2d` in the case of interest.

Does this look reasonable?
Also, shall we leave `cross2d` accessible only via `numba.targets.array_math_ext` or shall I import it in `targets/__init__.py` for instance?

Thank you in advance.

stuartarchibald commented:
Thanks for the above.

Perhaps wire it through a new `numba.numpy_extensions.cross2d`? @seibert IIRC you took at look at this first time round, how does that sound?

stuartarchibald commented:
> Thanks for the above.
> 
> Perhaps wire it through a new `numba.numpy_extensions.cross2d`? @seibert IIRC you took at look at this first time round, how does that sound?

Or in fact, just move the current `numba.targets.array_math_ext` to  `numba.numpy_extensions.cross2d`? I don't envisage there being a large number of these cases.

partmor commented:
> > Thanks for the above.
> > Perhaps wire it through a new `numba.numpy_extensions.cross2d`? @seibert IIRC you took at look at this first time round, how does that sound?
> 
> Or in fact, perhaps just move the current `numba.targets.array_math_ext` to `numba.numpy_extensions.cross2d`? I don't envisage there being a large number of these cases.

Yep, this is a lot better also IMO. I've proceeded to do this and updated documentation/references.

@stuartarchibald  Do you think it will be worth adding a line with `import .numpy_extensions` to `numba/__init__.py`?
Do yo usee anything else missing?

stuartarchibald commented on numba/tests/test_np_functions.py:
```diff
@@ -3065,6 +3066,78 @@ def test_cross_exceptions(self):
+                np.array((4, 5, 6))
+            )
+        self.assertIn(
+            'incompatible dimensions',
```
Probably best to check for "incompatible dimensions for 2D cross product" as the above is sufficiently generic it might be aliased.

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -4029,7 +4029,46 @@ def impl(a, b):
+        b_ = np.asarray(b)
+        if a_.shape[-1] != 2 or b_.shape[-1] != 2:
+            raise ValueError((
+                "incompatible dimensions for 2D cross product\n"
```
```suggestion
                "Incompatible dimensions for 2D cross product\n"
```

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -4029,7 +4029,46 @@ def impl(a, b):
+@generated_jit
+def cross2d(a, b):
+    if not type_can_asarray(a) or not type_can_asarray(b):
+        raise TypingError("Inputs must be array-like.")
```
This branch needs a test.

stuartarchibald commented on numba/numpy_extensions.py:
```diff
@@ -0,0 +1,10 @@
+"""
+Numpy extensions.
```
```suggestion
NumPy extensions.
```

stuartarchibald commented:
https://github.com/numba/numba/pull/4641 is a conflict free version, will merge that if CI passes before conflicts are fixed here.
